File: 131_PalindromePartitioning.py
##73.75%
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        def isPalindrome(left, right):
            while left < right:
                if s[left] != s[right]:
                    return False
                left += 1
                right -= 1
            return True
        
        def partitionHelper(left, right):
            if left >= right:
                res.append(vec[:])
            for i in xrange(left, right):
                if isPalindrome(left, i):
                    vec.append(s[left : i + 1])
                    partitionHelper(i + 1, right)
                    vec.pop()
        
        if not s:
            return list()
        if len(s) == 1:
            return [[s]]
        res = list()
        vec = []
        partitionHelper(0, len(s))
        return res

File: 131_PalindromePartitioning.cpp
//73.9%
class Solution {
public:
    vector<vector<string>> partition(string s) {
        if(s.empty())
			return vector<vector<string> > ();
		if(s.size() == 1)
			return vector<vector<string> >(1, vector<string>(1, s));
		vector<vector<string> > res;
		vector<string> vec;
		partitionHelper(res, vec, s, 0, s.size() - 1);
		return res;
    }
private:
	void partitionHelper(vector<vector<string> >& res, vector<string>& vec, const string& s, int left, int right){
		if(left > right){
			res.push_back(vec);
			return;
		}
		for(int i = left; i <= right; ++i){
			if(isPalindrome(s, left, i)){
				vec.push_back(s.substr(left, i - left + 1));
				partitionHelper(res, vec, s, i + 1, right);
				vec.pop_back();
			}
		}	
	}
	bool isPalindrome(const string& s, int left, int right){
		while(left < right){
			if(s[left++] != s[right--])
				return false;
		}
		return true;
	}
};